_posts/2017-11-22-the-evolution-of-a-python-script.md

```diff
@@ -0,0 +1,148 @@
+  return do_something_to(sys.argv[2)
+
+do_something_else(sys.argv[2)
+
+# output:
+# File executed: script.py
+# Parameter 1: param1
+# Parameter 2: param2
+```
+
+As you can see, python makes it very simple to pass values from the command line to the script. The script is able to act on different resources and react to the parameter values. The dependency between the authors computer configuration is no longer relevant to run the script successfully. Users are not required to edit the source code to change the values of what are now the parameters.
+
+_This is another reason to execute scripts from the command line since it is not possible to pass parameters from the double-click file execution._
+
+While passing parameters in this fashion is a big improvement over no parameters, scripts with parameters require authors to document how the script is to be used. Without a readme or some instructions, a new user will not know to pass parameters nor what values to use. Scripts with more than two or three inputs can quickly become confusing to a user. Trying to remember the order of parameters can also be hard to remember.
+
+A semi acceptable fix to this problem would be to inspect the parameters sent to the script, if any, and validate them. If they are invalid, return a message to the user that the parameters were not valid and provide an example of how the script should be used. **Congratulations, your script has evolved into a console application.**
+
+## A Better Console Application - Using a 3rd Party CLI Module
+
+`python geocode.py milepost 300 15N`
+
+The python community has created many useful modules that we all can `pip` or `conda` install and start using. Command line interface modules are no different. There are a [lot](https://pypi.python.org/pypi/clint/) of [command](http://click.pocoo.org/) [line](https://pypi.python.org/pypi/plac) [interface](http://docs.openstack.org/developer/cliff/) [modules](http://builtoncement.com/) because there are a lot of opinions on how they should work.
+
+AGRC has been using [docopt](http://docopt.org/) and really enjoying it. It is a lightweight module that uses POSIX style comments to define how your console application can be used. You get a CLI and documentation!
+
+```py
+'''
+Milepost geocoding!
+
+Usage:
+  geocode milepost (<milepost> <route>)
+'''
+
+from docopt import docopt
+
+
+if __name__ == '__main__':
+    arguments = docopt(__doc__)
+```
+
+The above documentation can be parsed by docopt to help a developer create functionality for the milepost command. If a user were to type
+
+```
+python geocode.py 123 15N
+```
+
+docopt would parse the input, realize it **does not** match any documented commands, and return the available commands/syntax/documentation to the user. The user will realize that their syntax is missing the `milepost` and correct their mistake. This is the functionality we suggested above but now the 3rd party module is taking care of all of the boring stuff.
```

Comment:
"...missing the `milepost` and correct..." -> "...missing the `milepost` command and correct..."